plutus-metatheory/src/VerifiedCompilation/UCSE.lagda.md

```diff
@@ -0,0 +1,63 @@
+---
+title: VerifiedCompilation.UCSE
+layout: page
+---
+
+# Common Subexpression Elimination Translation Phase
+```
+module VerifiedCompilation.UCSE where
+
+```
+## Imports
+
+```
+open import VerifiedCompilation.Equality using (DecEq; _≟_; decPointwise)
+open import VerifiedCompilation.UntypedViews using (Pred; isCase?; isApp?; isLambda?; isForce?; isBuiltin?; isConstr?; isDelay?; isTerm?; allTerms?; iscase; isapp; islambda; isforce; isbuiltin; isconstr; isterm; allterms; isdelay)
+open import VerifiedCompilation.UntypedTranslation using (Translation; translation?; Relation)
+open import Relation.Nullary.Product using (_×-dec_)
+open import Data.Product using (_,_)
+import Relation.Binary as Binary using (Decidable)
+open import Relation.Nullary using (Dec; yes; no; ¬_)
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+import Relation.Binary.PropositionalEquality as Eq
+open Eq using (_≡_; refl)
+open import Data.Empty using (⊥)
+open import Agda.Builtin.Maybe using (Maybe; just; nothing)
+open import Untyped.RenamingSubstitution using (_[_])
+```
+## Translation Relation
+
+This module is required to certify that an application of CSE doesn't break the
+semantics; we are explicitly not evaluating whether the particular choice of
+sub-expression was a "good" choice.
+
+As such, this Translation Relation primarily checks that substituting the expression
+back in would yield the original expression.
+
+```
+data UCSE : Relation where
+  cse : {X : Set} {x : Maybe X ⊢} {x' e : X ⊢}
+    → Translation UCSE (x [ e ]) x'
```

Comment:
I'm perhaps nitpicking, but this isn't sufficient in the general case. CSE promises to pull some common subexpressions out, but it doesn't promise to pull all of them out. I.e. `x'` may still start with the same lambda and `e` may be substituted partially in `x`.